### PR TITLE
Tracking platform dependencies

### PR DESCRIPTION
**Pull request by mthalman**
This is the proposed design for https://github.com/dotnet/core/issues/5646 within the https://github.com/dotnet/core/issues/5651 epic.

This is a cross-cutting proposal that impacts all product teams.  I've tried to include representatives across the board as reviewers but feel free to include others you feel are missing.

### CODE REVIEW

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+A key piece of metadata that gives context to the dependency is the "usage" field. This field is set to one of the well-known values that describes the scenario in which this dependency applies. Here are some examples of usage values:
+
+* default: Indicates that the dependency applies to a canonical app scenario (i.e. Hello World).  Note that this is specifically about a canonical app and not intended to be a description of required dependencies in the absolute sense. An example of this is libicu. While libicu is not required to run an app if it has been set to use invariant globalization, the default/canonical setting of a .NET app is that invariant globalization is set to false in which case libicu is necessary. This is why the term "default" is used rather than something like "required".
+* diagnostics: Indicates the dependency should be used in scenarios where diagnostic tools are being used such as with LTTng-UST.
+* httpsys: Indicates the dependency should be used for ASP.NET Core apps that are configured to use the HTTP.sys web server.
+* localization: Indicates the dependency should be used for localization/globalization scenarios such as with tzdata.
```
Comment (marek-safar) [inline accepted/2021/platform-dependencies/platform-dependencies.md:224]: I don't think this will scale and at the same time, it's not granular enough. I can imagine a model where each NuGet can describe its dependencies as more maintainable and scalable.

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+
+```console
+dotnet-deps platform remove debian.9
+```
+
+## Requirements
+
+### Goals
+
+* Common schema capable of describing both runtime and toolchain dependencies.
```
Comment (marek-safar) [inline accepted/2021/platform-dependencies/platform-dependencies.md:56]: What do runtime and toolchain mean in this context?

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+* Model that limits repetition to make maintenance easier.
+* File format that is machine-readable to allow for automated transformation into other output formats.
+* Ability to describe:
```
Comment (marek-safar) [inline accepted/2021/platform-dependencies/platform-dependencies.md:59]: Will this apply only to shared framework model or also bundled/self-contained model?

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+    * Dependencies from various granularities of .NET components such as shared frameworks and NuGet packages.
+    * Dependencies that exist outside of canonical scenarios, such as opting into specific features like diagnostics.
+* Workflow that .NET contributors can easily follow for maintaining the model.
+
+### Non-Goals
+
+* Any dependency that is included in the deployment of the application is outside the scope of this design. NuGet packages are an example of this. A .NET application's dependency on a .NET package, and any assets contained in those packages (managed, native, or otherwise), is explicitly included in the deployment of the application itself. Therefore, the operating environment is not required to be pre-configured with those specific assets; it'll get them naturally through the deployment of the application. However, a NuGet package may have its own platform dependency (e.g. a Linux package) that is not physically contained in the NuGet package; such a platform dependency would be in scope with this design. The concern addressed here is solely focused on what the operating environment must be pre-configured to contain in order to operate on .NET scenarios.
```
Comment (marek-safar) [inline accepted/2021/platform-dependencies/platform-dependencies.md:67]: I think this needs a bit word tweaking. You are not excluding NuGet package dependencies in the design as they are the core building block of any app and runtime itself ships many native libraries with native dependencies as OOB NuGets.

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+#### New Dependency
+
+New platform dependencies can be introduced by making a code change that has the dependency or including a new component into .NET (e.g. shared framework, NuGet package). When such a dependency is added, the contributor should also update the appropriate platform dependency model file with the necessary information that describes the dependency across all supported platforms.
+
+##### Change Detection
+
+In order to avoid the reliance upon contributors to recognize when they've changed a dependency, a more automated solution would be preferable. This can be done by defining a GitHub bot that checks for files in PRs containing `NativeMethods` or `Interop` in their name. If such a file is detected, a label is added to the PR alerting the submitter that they should evaluate their changes for changes to the dependencies. While not a fool-proof solution, it should provide coverage for the vast majority of dependencies. Work is still required by the submitter to make the appropriate changes to the platform dependency model but the GitHub bot helps alert them when there is potentially action that is needed.
```
Comment (jeffschwMSFT) [inline accepted/2021/platform-dependencies/platform-dependencies.md:354]: I don't think that this is sufficient.  To help keep dependency tracking live we will need to invest more deeply in dependency identification.  Also how would we ensure that toolchain dependencies are being identified automatically?

**Review by jeffschwMSFT — COMMENTED** [review]
IMO one of the reasons for having this information is to ensure that we are meeting all the constraints of our system.  For example, changes to minimal cmake versions need source build considerations.  Where in this model do we document why the min versions of all our dependencies are they way they are?  Having this will allow maintainers to more effectively understand how to change them/why they are they way they are.

**Comment by jkotas** [comment]
> For example, changes to minimal cmake versions need source build considerations. Where in this model do we document why the min versions of all our dependencies are they way they are? Having this will allow maintainers to more effectively understand how to change them/why they are they way they are.

+1

Raising minimum tool versions often requires weighting trade-offs between pain caused by staying with the old version and pain caused by requiring new minimum version. For example, we have different minimum cmake versions for different build configurations today because we found that aggressively raising the minimum version accross the board is going to incur too much pain in aggregate.

We should be able to learn how situations like this are handled in other ecosystems. For example, here is a recent discussion on raising minimum cmake version in LLVM: https://lists.llvm.org/pipermail/llvm-dev/2020-April/subject.html#140578

**Comment by Pilchie** [comment]
I'm not aware of any platform dependencies that ASP.NET Core has here, with the possible exception of the Http.sys/IIS components in Windows.

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+            }
+          ]
+        }
+      ]
+    }
+  ]
+}
+```
+
+The model above uses the Debian Linux distro as an example. It shows the `Microsoft.NETCore.App` shared framework as having a base set of dependencies but for Debian 10, the `libicu57` package dependency is overriden to be `libicu63` since that is the version available in Debian 10. It also shows that the `System.DirectoryServices.Protocols` NuGet package has a dependency on `libldap-2.4-2`.
```
Comment (danmoseley) [inline accepted/2021/platform-dependencies/platform-dependencies.md:325]: libicu is tricky because we will load any version of icu found, [as long as it's at least 50](https://github.com/dotnet/runtime/blob/0e8efce08c99fb4b3ee0d46839dbe4a05c996b84/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L309). We start probing for the nonexistent libicu80 and work down. We may know what version of libicu is on each target distro, but we will load any newer one they've added - does that matter? Does that need to be expressed? I'm guessing not, since the test is - will the app run.

If there was a distro we didn't track, that had a version of libicu that was not the default on any of our target distros, we would still load it, but this json file would consider the dependency to be missing. I assume that's OK as the json is only designed to cover supported distros.

Comment (danmoseley) [reply accepted/2021/platform-dependencies/platform-dependencies.md:325]: Flipping on its head, if libicu63 is always present on Debian10, what value does it add to list it explicitly as a dependency, rather than just expressing that Debian10 will satisfy our dependencies? Will someone be constructing a docker container that has only part of Debian?

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 5 above)
Comment (danmoseley) [reply accepted/2021/platform-dependencies/platform-dependencies.md:354]: I could imagine something like - run all the tests under `ltrace` or similar, and compare all the dlopen's seen to a baseline list. When a new one is seen, it's an indication that someone should check whether a json update is needed, then update the baseline.

It seems to me that isn't part of 'crawl' but maybe 'run'

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 6 above)
Comment (danmoseley) [reply accepted/2021/platform-dependencies/platform-dependencies.md:325]: Perhaps what I really want to know is, given supported distro XYZ, what packages do I need to explicitly add on top?

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+
+* `rid`: rid of the platform to remove
+
+Example:
+
+Removes the debian.10 platform:
+
+```console
+dotnet-deps platform remove debian.10
+```
```
[the remark this answers is not in the document]
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:433]: I've looked at Node.js and Python so far and have not come across anything formalized like this.

@omajid - Have you or anyone else at RedHat come across any other platforms that do anything similar to this?

@tianon - In your role maintaining Docker Hub's official images, have there been any platforms out there that have created a standalone, machine-readable description of the platform's package dependencies such that it can be used to maintain the packages installed by the Dockerfile?  Or has it always just been the Dockerfile being the source of truth?  To give you a very brief summary of what is being proposed here for .NET, there would be a JSON file which formalizes .NET's platform dependencies, including Linux packages.  That JSON file could be used to do transformations into or validation of Dockerfiles, as an example.  The intention is to have a standalone, independent source of truth from which other assets could be derived.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 3 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:59]: Are you referring to the framework-dependent/self-contained deployment models?  If so, this applies to both because in either model the platform dependencies are still required.  Self-contained deployment does not statically link these dependencies.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 6 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:325]: Did you mean to mark this conversation as resolved?

> We may know what version of libicu is on each target distro, but we will load any newer one they've added - does that matter? Does that need to be expressed?

One answer to this is that we're currently [documenting](https://docs.microsoft.com/en-us/dotnet/core/install/linux-debian#dependencies) the required libicu version; we'd want to ensure that is accurate and reflects the current state of the distro's repository.  I do see there being value in having an explicit package name in general, because it obviates the need for there to be any special interpretation.  If this model is going to be used by the dotnet-install script, we don't want to have it just say "you need libicu, whatever version is fine"; you'd want it to be an explicit package name.  This makes things simple for the consuming perspective.

> Flipping on its head, if libicu63 is always present on Debian10, what value does it add to list it explicitly as a dependency, rather than just expressing that Debian10 will satisfy our dependencies? Will someone be constructing a docker container that has only part of Debian?

First, libicu is not present by default in Debian 10, at least for the Docker image.  Again, by being explicit about the dependencies you remove interpretation.  Let's say Debian 10 did include libcu.  Does the lack of including it in our model mean that we don't require or that we do but it's there by default?  That lack of information makes it difficult to do analysis of the dependencies; for example when needing to determine what the dependencies are to respond to a security incident as I've described [here](https://github.com/dotnet/designs/pull/184/files#diff-203b0b60bf87544ba5e19865ea38a42b251f78c4a904756e7499d768bb5f59d0R34).  There are other scenarios as well.  Let's say that all of the distros we support have libicu by default and so we don't describe libicu in the model as a dependency.  Now we want to evaluate another distro for inclusion in the  supported list.  Does it have the dependencies we need?  We don't know because we haven't described them fully.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 1 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:213]: I've made an update so that these usages are self-described within the model: https://github.com/dotnet/designs/pull/184/commits/8c52259d25a6610d8c8a1c6f228d303210575c4f.  This adds a mapping at the root of the model to define all the available dependency usages so that there can still be model validation and a way to provide a description of the value.  This will allow teams to add more usages as they need and confine the updates to just the model file.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 2 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:61]: Added a definitions section which clarifies these terms: https://github.com/dotnet/designs/pull/184/commits/e38819540b74f3df2ecb087a3fb34c13fcbf4937

**Comment by knuxbbs** [comment]
Is this proposal related to workload manifests (#120)?

For instance, how to track platform dependencies for a mobile workload?

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (tianon) [reply accepted/2021/platform-dependencies/platform-dependencies.md:433]: This is really ambitious -- I'm not aware of anyone trying to standardize this sort of metadata.

If I'm understanding correctly, you're trying to come up with a standard identifier for things like `libargon2-dev` in Debian/Ubuntu vs `argon2-dev` in Alpine etc?

I think a really big challenge you'll likely run into is that these things aren't always one-to-one mapped -- there are many cases where distributions provide multiple variants (see the `-dev` packages in https://packages.debian.org/source/sid/curl for example) or even split different binaries across packages differently.  In addition, these things tend to move between packages over time too (like how Debian's `btrfs-progs` used to also contain the development headers until the dedicated `libbtrfs-dev` package was introduced).  Even the "upstream project" often gets split differently in different distros (where some have a single source package and others will end up with multiple source packages representing the same thing).

A different approach to trying to identify/recognize all the explicit packages might be to identify commands and specific header/`pkg-config` files necessary, but even that's going to be a bit of guesswork (and if it isn't automated to some extent, will struggle with bitrot), and that list isn't always straightforward to come up with, even for a human (like how difficult it would be to determine in an automated way that a program ends up invoking `git` at runtime, or that it uses `dlopen` to load a shared library dynamically).

There's also going to be different classes of dependencies -- obvious ones are build vs runtime, but even at runtime there are some dependencies that are more "required" than others (see https://www.debian.org/doc/debian-policy/ch-relationships.html for an example of how Debian handles this with `Depends` vs `Recommends` vs `Suggests` etc).

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+
+## Requirements
+
+### Goals
+
+* Common schema capable of describing both runtime and toolchain dependencies.
+* Model that limits repetition to make maintenance easier.
+* File format that is machine-readable to allow for automated transformation into other output formats.
+* Ability to describe:
+    * Dependencies for multiple platform types (Linux, Windows, MacOS)
```
Comment (omajid) [inline accepted/2021/platform-dependencies/platform-dependencies.md:65]: What about more specifics in that platform? 

For example, .NET Core 2.0 only supported using OpenSSL 1.0.x. .NET Core 2.1 got additional fixes that let it also use OpenSSL 1.1.y. It would be great it we could express things like that in this manifest: `needs-one-of (OpenSSL 1.0, OpenSSL 1.1)`.

I am also wondering what happens across builds. A portable build of .NET 5 SDK that bundles dependencies will have fewer build and runtime requirements than a non-portable build of .NET 5 SDK. The non-portable build will, for example, need the matching version of OpenSSL.

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+        public IList<PlatformDependency> Dependencies { get; }
+    }
+
+    // Types of .NET components
+    public enum ComponentType
+    {
+        // A shared framework (e.g. Microsoft.NETCore.App, Microsoft.AspNetCore.App)
+        SharedFramework,
+
+        // A NuGet package (e.g. System.Drawing.Common)
```
Comment (omajid) [inline accepted/2021/platform-dependencies/platform-dependencies.md:150]: Since you are calling out System.Drawing.Common here, I can't help but wonder what that will look like. AFAIAA, `System.Drawing.Common` needs `libgdiplus`, but `libgdiplus` is neither a build-time nor a runtime dependency, unless you are explicitly using `System.Drawing.Common` at runtime. Does it fit in with the `default` type or does this need something else?

Comment (danmoseley) [reply accepted/2021/platform-dependencies/platform-dependencies.md:150]: @safern

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+
+The platform dependency model describes dependencies by name and those names are subject to change from upstream sources.
+
+> A classic example of this is the [ICU library](http://site.icu-project.org/). In many Linux distros, the ICU library's package contains the version number in the name (e.g. libicu57). When a Linux distro updates to use a new version of the ICU library, they typically remove the previous version from the package repository. Thus, any dependency described for the ICU library needs to account for the name change.
+
+We will rely on manual discovery to determine when the name of a package has changed since the typical scenario in which a name change is detected happens when a build breaks. When a dependency name change is discovered, the person discovering the change should file an issue in dotnet/core as described [above](#maintenance-scenarios).
+
+#### New Platform Support
+
+When support for a new platform is added to the product, the platform dependency model of future releases needs to be updated include this platform and all its supported versions.
```
Comment (omajid) [inline accepted/2021/platform-dependencies/platform-dependencies.md:366]: Is there any impact on unsupported platforms? For example, we are trying to get source-build to work with Arch Linux. If those folks get .NET building on Arch, will they be affected if we start tracking dependencies here? Will anything suddenly start breaking for them? And will it be okay if we don't want to track Arch Linux here if we dont want to "officially support" it?

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+
+        // Type of the referenced dependency
+        public DependencyType DependencyType { get; set; }
+    }
+}
+```
+
+#### Platform
+
+Each platform has its own way of identifying assets that .NET is dependent upon. For example, the Open SSL library is labeled as openssl-libs in Fedora and libssl in Debian.  For that reason, platform dependencies can't be described without identifying which platform they are for. The model handles this by describing platforms as a top-level concept to organize dependencies. Since many dependencies apply to multiple versions of a given platform, the ability to describe a versionless platform, as well as more specific versioned platforms if necessary, is possible. This is done by associating a platform with a rid. Rids are hierarchical allowing more specific platforms to be targeted which provides an ideal mechanism to be able to flexibly describe the scope of dependencies. In the same way that rids are hierarchical, so too are the platforms in this model. This allows for less repetition in the actual model content and enables platforms with more specificity to append or amend dependencies from their parent platform. Some examples of this:
```
Comment (omajid) [inline accepted/2021/platform-dependencies/platform-dependencies.md:200]: In practice, I have seen a couple of issues in how RIDs are used that doesn't quite match this. In particular, RIDs are not scalable:  https://github.com/NuGet/Home/issues/5862

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (omajid) [reply accepted/2021/platform-dependencies/platform-dependencies.md:433]: > Have you or anyone else at RedHat come across any other platforms that do anything similar to this?

I can try asking around, but I have not seen anything similar to this. Most upstreams are less disciplined than this, I think :smile:

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 9 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:150]: > unless you are explicitly using `System.Drawing.Common` at runtime

That's the key.  The dependency would specifically be tied to the `System.Drawing.Common` NuGet package.  So if you're referencing `System.Drawing.Common` at runtime, then you'll require `libgdiplus` for the canonical scenario.  In that case, the type would be `default`.

Here's a snippet of what that would look like:

```json
{
  "platforms":
  {
    "rid": "debian",
    "components": [
      {
        "name": "System.Drawing.Common",
        "type": "NuGetPackage",
        "platformDependencies": [
          {
            "name": "libgdiplus",
            "dependencyType": "LinuxPackage",
            "usage": "default"
          }
        ]
      }
    ]
  }
}
```

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 11 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:200]: Are there other issues besides the scalability problem?

I don't see the lack of RID scalability being an issue here.  I still think RIDs are the appropriate identifier to use here because it conceptually fits.  I don't want to be inventing a new hierarchy system.  And if it has the same limitations as the NuGet domain, to me that's fine; at least it's consistent.  I don't see this as a problem that this design should solve when NuGet itself hasn't solved it.

Comment (omajid) [reply accepted/2021/platform-dependencies/platform-dependencies.md:200]: Fair enough. I wanted to call it out because we are actively trying to find a fix/solution for this.

I am not aware of any non-scalability problems.

Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:200]: Ok, thanks for calling it out as I wasn't aware of it.  I'll keep my eyes on it so that any changes there could potentially be incorporated here.

File: accepted/2021/platform-dependencies/platform-dependencies.md [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+
+#### Component
+
+Within each platform can be described a set of components. These components represent logical parts of .NET that are delivered as a unit. Examples of these are shared frameworks (Microsoft.NETCore.App, Microsoft.AspNetCore.App, etc) and NuGet packages.
+
+#### Platform Dependency
+
+Each component describes the dependencies it has for the platform it is contained within. A dependency is identified by its name and type (Linux distro package, DLL).
+
+A key piece of metadata that gives context to the dependency is the "usage" field. This field is set to one of the well-known values that describes the scenario in which this dependency applies. Here are some examples of usage values:
```
[the remark this answers is not in the document]
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:217]: I made an update to address this, including an example: https://github.com/dotnet/designs/pull/184/commits/8b0ced74745e483cc8d211de68eb314646d517de.  Hopefully that makes sense.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:484]: Thanks, @tianon

> If I'm understanding correctly, you're trying to come up with a standard identifier for things like `libargon2-dev` in Debian/Ubuntu vs `argon2-dev` in Alpine etc?

Not really a standard identifier.  There's no attempt to map 1:1 between distros or anything like that.  Package dependencies can be defined for each distro independently.  It's really no different than the dependencies that are described in [documentation form](https://docs.microsoft.com/dotnet/core/install/linux), it's just that they're defined in a machine-readable form using a common schema.

> There's also going to be different classes of dependencies -- obvious ones are build vs runtime

Yes, that's [accounted for](https://github.com/dotnet/designs/pull/184/files#diff-203b0b60bf87544ba5e19865ea38a42b251f78c4a904756e7499d768bb5f59d0R22-R23) here by having separate models between them.

>  but even at runtime there are some dependencies that are more "required" than others (see https://www.debian.org/doc/debian-policy/ch-relationships.html for an example of how Debian handles this with Depends vs Recommends vs Suggests etc).

This is sort of addressed by the [dependency usage](https://github.com/dotnet/designs/pull/184/files#diff-203b0b60bf87544ba5e19865ea38a42b251f78c4a904756e7499d768bb5f59d0R108-R111) concept. It's less prescriptive than the depends/recommends/suggests model, allowing for the dependency to be associated with specific dev or app scenarios.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 5 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:405]: This is probably the aspect of this proposal that will require the most iteration and refinement over time.  I don't want to over-invest in this, however. I fully admit that the proposed change detection isn't foolproof.  But it's a fairly cheap and non-disruptive solution to get started with.  I think as we learn of other areas that require detection, we can work to address those as needed.

I do like the suggestion of @danmoseley for an advanced implementation of detection logic.  This could apply to both runtime and toolchain dependencies.

**Comment by mthalman** [comment]
> > For example, changes to minimal cmake versions need source build considerations. Where in this model do we document why the min versions of all our dependencies are they way they are? Having this will allow maintainers to more effectively understand how to change them/why they are they way they are.
> 
> +1
> 
> Raising minimum tool versions often requires weighting trade-offs between pain caused by staying with the old version and pain caused by requiring new minimum version. For example, we have different minimum cmake versions for different build configurations today because we found that aggressively raising the minimum version accross the board is going to incur too much pain in aggregate.
> 
> We should be able to learn how situations like this are handled in other ecosystems. For example, here is a recent discussion on raising minimum cmake version in LLVM: https://lists.llvm.org/pipermail/llvm-dev/2020-April/subject.html#140578

@jkotas, @jeffschwMSFT - I'm going to throw out some options here.  Let me know what you think is reasonable or feel free to suggest alternatives.  My assumption here is that any sort of description will need to be free-form text, not fitting to any schema.

#### 1. Include the description within the model JSON file as general purpose "dependency notes"

Pros:
* Notes are versioned together with the model file.
* Edits are reviewable via PRs

Cons:
* Readability and maintainability suffers by working with free-form text in a file format that is primarily intended for machine-readability.

#### 2. Define description in a GitHub issue with the URL referenced by the dependency in the model.

Pros:
* Provides the benefits of GitHub issues (e.g. conversation-style timeline, people tagging, etc).

Cons:
* Not versioned.  Are there separate issues between different .NET releases for the same dependency?
* Edits are not reviewable via PRs

#### 3. Define description in a Markdown file in the relevant GitHub repo with the commit URL referenced by the dependency in the model.

Pros:
* Versionable, provides history over time
* Edits are reviewable via PRs

Cons:
* Versions independently of the model file. Requires updating the model's commit URL reference with any change to the Markdown file.

**Comment by jkotas** [comment]
My preference would be combination of 1 and 2. I think it is useful to have space for a short free-form comment in the model file. If there is too much to say, this comment can include link to github issues, documentation, etc.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 5 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:405]: It needs to be complete to the extent that we're aware of the dependencies.  The model is intended to be used as input to other dependent artifacts like Dockerfiles, as you suggest (see phases 2 & 4 in https://github.com/dotnet/core/issues/5651 for others).  I don't see the ability to know and describe the dependencies as being unfeasible.  Indeed, it better not be; otherwise, what are we documenting for customers?  The change detection aspect is certainly a challenge and I feel should be scoped, at least to begin with, to be able to detect most dependency changes rather than all.

My view is that we treat the dependency model as a work in progress for the lifetime of the .NET version it's associated with.  We do our best effort to define the model and if we discover something later on several releases later, we can just go back and edit it to accurately reflect what is known.  The benefit of having downstream assets consuming this model data is that it acts a form of validation.  If the Dockerfiles are synced with the dependency model and something doesn't work in the container because it's missing a dependency, then we know the model isn't correct.  I think we just keep iterating on it until it's right.

Comment (danmoseley) [reply accepted/2021/platform-dependencies/platform-dependencies.md:405]: Ah, somehow I had thought you were trying to describe the closure, ie., including transitive dependencies. Yes, just describing direct dependencies is surely doable.

**Comment by mthalman** [comment]
@knuxbbs -

> Is this proposal related to workload manifests (#120)?

It's unrelated to #120.  Workloads are a set of references to SDK packs and workload manifests are descriptions of those workloads.  Workarounds are not concerned with, nor do they provide, the set of platform dependencies necessary to run in the target environment.

> For instance, how to track platform dependencies for a mobile workload?

I'll rephrase this as "how to track platform dependencies for a mobile **platform**".  The schema supports this by the use of a RID to describe the platform.  In the case of mobile platforms, there are RIDs for [Android](https://github.com/dotnet/runtime/blob/1f213adf3ccc0a4632049ff8b32bc493337578b4/src/libraries/Microsoft.NETCore.Platforms/pkg/runtime.json#L210) and [iOS](https://github.com/dotnet/runtime/blob/1f213adf3ccc0a4632049ff8b32bc493337578b4/src/libraries/Microsoft.NETCore.Platforms/pkg/runtime.json#L1088), for example.  So it would just be another platform described in the model.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 3 above)
Comment (marek-safar) [reply accepted/2021/platform-dependencies/platform-dependencies.md:64]: > Self-contained deployment does not statically link these dependencies.

That's not correct, some of them do.

Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:64]: Can you provide more details?  I'd like to know more about that.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 10 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:417]: No, nothing would break.  This is purely a representation of what dependencies exist; it doesn't dictate what those dependencies must be.  The flow of information would always be that the maintainers of the .NET source define what dependencies they want to have, and then the dependency model gets updated to reflect that.  There would be no use of the dependency model at runtime or build time (i.e. it has no impact on the running of a .NET application or of the building of .NET source). Its purpose is to add rigor and correctness to the maintenance of other assets that describe what the dependencies are (such as documentation).

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 4 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md]: I've reworded this to be more accurate: https://github.com/dotnet/designs/pull/184/commits/1a925925642a9bce2c580498445bf0896a522740

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 8 above)
Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:65]: > For example, .NET Core 2.0 only supported using OpenSSL 1.0.x. .NET Core 2.1 got additional fixes that let it also use OpenSSL 1.1.y. It would be great it we could express things like that in this manifest: needs-one-of (OpenSSL 1.0, OpenSSL 1.1).

I've redesigned the schema to support this.  Dependency names can now be described as an expression with logical OR operators.  Take a look: https://github.com/dotnet/designs/pull/184/commits/b7833d0705cfcaa9bc23f5d3024099056c492fe0

> I am also wondering what happens across builds. A portable build of .NET 5 SDK that bundles dependencies will have fewer build and runtime requirements than a non-portable build of .NET 5 SDK. The non-portable build will, for example, need the matching version of OpenSSL.

Since that would have been done by a third party and customized to suit their desired configuration, it's not really possible to describe that here, nor is it really relevant. The intent is to describe the dependencies of the assets distributed by Microsoft.  And as I mention [here](https://github.com/dotnet/designs/pull/184#discussion_r596927235), the model description has no impact on the functionality of .NET that has been built/distributed external of Microsoft.

File: accepted/2021/platform-dependencies/platform-dependencies.md (hunk 3 above)
Comment (marek-safar) [reply accepted/2021/platform-dependencies/platform-dependencies.md:64]: For example, Blazor is a self-contained deployment setup with only statically linked dependencies.

Comment (mthalman) [reply accepted/2021/platform-dependencies/platform-dependencies.md:64]: Ah, thanks for clarifying. In that case, the `browser` or `browser-wasm` RIDs would be just another platform described in this model and not contain these dependencies or whatever would be appropriate.

**Comment by mthalman** [comment]
All current feedback has been addressed. This is ready for further review.
